File: services/tavrion-bot-api/live_fetch.py
"""Live page fetch for price/time-sensitive bot queries."""

from __future__ import annotations

import re
from urllib.parse import urljoin, urlparse

import httpx

LIVE_QUERY_RE = re.compile(
    r"\b(price|pricing|cost|fee|fees|plan|plans|subscription|subscribe|how much|"
    r"current|latest|today|now|live|updated|rate|rates|tier|tiers|package|packages|"
    r"quote|billing|per month|/mo|annual)\b",
    re.I,
)
PRICING_URL_RE = re.compile(r"pricing|price|plans?|subscribe|billing|cost", re.I)


def needs_live_fetch(message: str) -> bool:
    return bool(LIVE_QUERY_RE.search(message))


def _strip_html(html: str) -> str:
    html = re.sub(r"<script[\s\S]*?</script>", " ", html, flags=re.I)
    html = re.sub(r"<style[\s\S]*?</style>", " ", html, flags=re.I)
    html = re.sub(r"<[^>]+>", " ", html)
    return re.sub(r"\s+", " ", html).strip()


def _extract_meta(html: str, name: str) -> str:
    escaped = re.escape(name)
    patterns = [
        rf'<meta[^>]*(?:name|property)=["\']{escaped}["\'][^>]*content=["\']([^"\']*)["\']',
        rf'<meta[^>]*content=["\']([^"\']*)["\'][^>]*(?:name|property)=["\']{escaped}["\']',
    ]
    for pat in patterns:
        m = re.search(pat, html, re.I)
        if m and m.group(1).strip():
            return m.group(1).strip()
    return ""


async def fetch_live_page_content(url: str) -> dict | None:
    try:
        async with httpx.AsyncClient(follow_redirects=True, timeout=15.0) as client:
            res = await client.get(url, headers={"User-Agent": "TavrionBot-LiveFetch/1.0"})
        if res.status_code >= 400:
            return None
        html = res.text
        title = (
            _extract_meta(html, "og:title")
            or (re.search(r"<title[^>]*>([\s\S]*?)</title>", html, re.I) or [None, ""])[1]
        )
        title = re.sub(r"<[^>]+>", "", title or "").strip() or url
        description = _extract_meta(html, "og:description") or _extract_meta(html, "description")
        body = _strip_html(html)[:8000]
        content = "\n\n".join(p for p in [title, description, body] if p)
        if len(content) < 40:
            return None
        return {"url": str(res.url), "title": title, "content": content}
    except Exception:
        return None


def rank_urls_for_live_fetch(
    message: str,
    chunk_urls: list[dict],
    source_url: str,
) -> list[str]:
    seen: set[str] = set()
    ranked: list[tuple[str, int]] = []

    def add(url: str, score: int) -> None:
        if not url or url in seen:
            return
        try:
            parsed = urlparse(url)
            if parsed.scheme not in ("http", "https"):
                return
            seen.add(url)
            ranked.append((url, score))
        except Exception:
            pass

    add(source_url, 1)
    for c in chunk_urls:
        url = c.get("url", "")
        title = c.get("title", "")
        score = 2
        if PRICING_URL_RE.search(url):
            score += 10
        if PRICING_URL_RE.search(title):
            score += 5
        if needs_live_fetch(message) and PRICING_URL_RE.search(f"{url} {title}"):
            score += 8
        add(url, score)

    if needs_live_fetch(message):
        try:
            base = urlparse(source_url)
            for path in ("/pricing", "/plans", "/price", "/subscribe", "/billing"):
                add(f"{base.scheme}://{base.netloc}{path}", 12)
        except Exception:
            pass

    ranked.sort(key=lambda x: x[1], reverse=True)
    return [u for u, _ in ranked[:3]]


async def build_live_context(urls: list[str]) -> tuple[str, list[dict]]:
    import asyncio

    results = await asyncio.gather(*(fetch_live_page_content(u) for u in urls))
    valid = [r for r in results if r]
    if not valid:
        return "", []

    text = "\n\n".join(
        f"[LIVE {i + 1}] {p['title']} ({p['url']}) — fetched just now:\n{p['content']}"
        for i, p in enumerate(valid)
    )
    sources = [{"url": p["url"], "title": p["title"], "live": True} for p in valid]
    return text, sources


def extract_urls_from_text(text: str) -> list[str]:
    urls: list[str] = []
    seen: set[str] = set()
    for m in re.finditer(r"https?://[^\s<>\"')\]]+", text, re.I):
        u = m.group(0).rstrip(".,;:!?)")
        if u not in seen:
            seen.add(u)
            urls.append(u)
        if len(urls) >= 5:
            break
    return urls

File: supabase/functions/_shared/liveFetch.ts
const LIVE_QUERY_RE = /\b(price|pricing|cost|fee|fees|plan|plans|subscription|subscribe|how much|current|latest|today|now|live|updated|rate|rates|tier|tiers|package|packages|quote|billing|per month|\/mo|annual)\b/i;

const PRICING_URL_RE = /pricing|price|plans?|subscribe|billing|cost/i;

export function needsLiveFetch(message: string): boolean {
  return LIVE_QUERY_RE.test(message);
}

function stripHtmlToText(html: string): string {
  return html
    .replace(/<script[\s\S]*?<\/script>/gi, " ")
    .replace(/<style[\s\S]*?<\/style>/gi, " ")
    .replace(/<[^>]+>/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function extractMeta(html: string, name: string): string {
  const escaped = name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const patterns = [
    new RegExp(`<meta[^>]*(?:name|property)=["']${escaped}["'][^>]*content=["']([^"']*)["']`, "i"),
    new RegExp(`<meta[^>]*content=["']([^"']*)["'][^>]*(?:name|property)=["']${escaped}["']`, "i"),
  ];
  for (const re of patterns) {
    const m = re.exec(html);
    if (m?.[1]?.trim()) return m[1].trim();
  }
  return "";
}

export async function fetchLivePageContent(url: string): Promise<{ url: string; title: string; content: string } | null> {
  try {
    const res = await fetch(url, {
      headers: { "User-Agent": "TavrionBot-LiveFetch/1.0 (+https://jointavrion.com)" },
      redirect: "follow",
    });
    if (!res.ok) return null;
    const html = await res.text();
    const title = extractMeta(html, "og:title")
      || html.match(/<title[^>]*>([\s\S]*?)<\/title>/i)?.[1]?.replace(/<[^>]+>/g, "").trim()
      || url;
    const description = extractMeta(html, "og:description") || extractMeta(html, "description");
    const body = stripHtmlToText(html).slice(0, 8000);
    const content = [title, description, body].filter(Boolean).join("\n\n");
    if (content.length < 40) return null;
    return { url: res.url || url, title, content };
  } catch {
    return null;
  }
}

export function rankUrlsForLiveFetch(
  message: string,
  chunkUrls: { url: string; title?: string }[],
  sourceUrl: string,
): string[] {
  const seen = new Set<string>();
  const ranked: { url: string; score: number }[] = [];

  const add = (url: string, score: number) => {
    if (!url || seen.has(url)) return;
    try {
      const u = new URL(url);
      if (!["http:", "https:"].includes(u.protocol)) return;
      seen.add(url);
      ranked.push({ url, score });
    } catch { /* skip */ }
  };

  add(sourceUrl, 1);

  for (const c of chunkUrls) {
    let score = 2;
    if (PRICING_URL_RE.test(c.url)) score += 10;
    if (PRICING_URL_RE.test(c.title || "")) score += 5;
    if (needsLiveFetch(message) && PRICING_URL_RE.test(`${c.url} ${c.title}`)) score += 8;
    add(c.url, score);
  }

  if (needsLiveFetch(message)) {
    try {
      const base = new URL(sourceUrl);
      for (const path of ["/pricing", "/plans", "/price", "/subscribe", "/billing"]) {
        add(`${base.origin}${path}`, 12);
      }
    } catch { /* skip */ }
  }

  return ranked.sort((a, b) => b.score - a.score).slice(0, 3).map((r) => r.url);
}

export async function buildLiveContext(urls: string[]): Promise<{ text: string; sources: { url: string; title: string }[] }> {
  const results = await Promise.all(urls.map((u) => fetchLivePageContent(u)));
  const valid = results.filter((r): r is NonNullable<typeof r> => r !== null);
  if (!valid.length) return { text: "", sources: [] };

  const text = valid.map((p, i) =>
    `[LIVE ${i + 1}] ${p.title} (${p.url}) — fetched just now:\n${p.content}`
  ).join("\n\n");

  return {
    text,
    sources: valid.map((p) => ({ url: p.url, title: p.title })),
  };
}

export function extractUrlsFromText(text: string): string[] {
  const urls = new Set<string>();
  for (const m of text.matchAll(/https?:\/\/[^\s<>"')\]]+/gi)) {
    urls.add(m[0].replace(/[.,;:!?)]+$/, ""));
    if (urls.size >= 5) break;
  }
  return [...urls];
}
